In [1]:
#Importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Defining path
path = r'/Users/richardniyongabo/Desktop/Tasks_4.6_richard'

In [3]:
#Import data 1
ords_prods_merged = pd.read_pickle(os.path.join(path, 'orders_products_merged_ver2.pkl'))


In [4]:
df = ords_prods_merged

# Aggregating data

In [5]:
# Group then aggregate 
df.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.213779
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170828
12               15.887622
13               16.583304
14               16.757377
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# maximum average order number is 22.902379 and minimum average order is 15.213779

# Create a loyalty flag

In [7]:
# creating a new column for loyalty
ords_prods_merged['max_order'] = ords_prods_merged.groupby(['user_id'])['order_number'].transform(np.max)

In [8]:
#remove maximum number of roles for display
pd.options.display.max_rows = None

In [9]:
ords_prods_merged.head(100)

order_id  product_id  add_to_cart_order  reordered  Unnamed: 0  user_id  \
0          2       33120                  1          1     3355525   202279   
1         26       33120                  5          0     2549350   153404   
2        120       33120                 13          0      395303    23750   
3        327       33120                  5          1      976751    58707   
4        390       33120                 28          1     2764474   166654   
5        537       33120                  2          1     2985383   180135   
6        582       33120                  7          1     3204665   193223   
7        608       33120                  5          1     1514503    91030   
8        623       33120                  1          1      627389    37804   
9        689       33120                  4          1     1813801   108932   
10       726       33120                  3          1      956084    57362   
11       771       33120                 29          1     1230777    73968   
12       800       33120                  3          1     2205942   132856   
13       901       33120                  3          1     2669006   160657   
14      1005       33120                  5          1     1811669   108796   
15      1097       33120                 14          1      505328    30502   
16      1193       33120                  5          1     1897452   113928   
17      1300       33120                  1          1      488234    29456   
18      1887       33120                  6          1      416523    25085   
19      2082       33120                 20          0     1803679   108338   
20      2357       33120                  6          1     1118747    67308   
21      2551       33120                  6          1      346208    20944   
22      2652       33120                  1          1      623084    37557   
23      2846       33120                 10          1     2786191   168005   
24      2929       33120                  1          0     2868514   173056   
25      3031       33120                  2          1       41515     2555   
26      3302       33120                  3          0     1947835   116952   
27      3304       33120                  4          1      707091    42591   
28      3351       33120                  3          1     1101558    66340   
29      3365       33120                  2          1     1875065   112555   
30      3426       33120                  3          1     2829331   170703   
31      3488       33120                 21          1     1109922    66781   
32      4075       33120                 12          1     2107070   126715   
33      4105       33120                  5          1      428881    25781   
34      4254       33120                 10          1     2147551   129196   
35      4560       33120                  2          0     2334887   140576   
36      4731       33120                  4          0     1471172    88415   
37      5256       33120                 10          1     2702466   162759   
38      5640       33120                  9          1        2417      155   
39      5733       33120                  8          0     1534750    92179   
40      5812       33120                  4          0     1307165    78567   
41      5864       33120                  2          0      942071    56509   
42      6173       33120                  1          1     2588810   155845   
43      6250       33120                  1          1     2453772   147684   
44      6365       33120                  4          0     1667352   100101   
45      6675       33120                 16          1     3121029   188156   
46      6874       33120                  4          1      665541    40092   
47      6937       33120                  5          1     1399748    84096   
48      7267       33120                  3          1     1891705   113565   
49      7518       33120                  6          1

In [10]:
# use the loc() function to group the customers
ords_prods_merged.loc[ords_prods_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merged.loc[(ords_prods_merged['max_order'] <= 40) & (ords_prods_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merged.loc[ords_prods_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [12]:
ords_prods_merged['loyalty_flag'].value_counts(dropna = False)

Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: loyalty_flag, dtype: int64

In [13]:
#checking the average price of product purchase by the customer groups
ords_prods_merged.groupby('loyalty_flag').agg({'prices':['mean','max','min']})

prices              
                       mean      max  min
loyalty_flag                             
Loyal customer    10.388747  99999.0  1.0
New customer      13.294370  99999.0  1.0
Regular customer  12.496203  99999.0  1.0

# prices of products differ for loyal customer (low spending habits), new customer (high spending habits) and followed by regular customer

# Spender Flag 

In [14]:
#creating a new column for average prices for each user
ords_prods_merged['avg_price'] = ords_prods_merged.groupby(['user_id'])['prices'].transform(np.mean)

In [15]:
#derive new column with loc
ords_prods_merged.loc[ords_prods_merged['avg_price'] >= 10, 'spender_flag'] = 'High spender'
ords_prods_merged.loc[ords_prods_merged['avg_price'] < 10, 'spender_flag'] = 'Low spender'

In [16]:
ords_prods_merged['spender_flag'].value_counts(dropna = False)

Low spender     31798751
High spender      635461
Name: spender_flag, dtype: int64

# Frequency Flag 

In [17]:
#creating a new column according to median since last order
ords_prods_merged['median_order_duration'] = ords_prods_merged.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [18]:
#derive new column with loc
ords_prods_merged.loc[ords_prods_merged['median_order_duration'] > 20, 'frequency_flag'] = 'Non-frequent customer'
ords_prods_merged.loc[(ords_prods_merged['median_order_duration'] > 10) & (ords_prods_merged['median_order_duration'] <= 20), 'frequency_flag'] = 'Regular customer'
ords_prods_merged.loc[ords_prods_merged['median_order_duration'] < 10, 'frequency_flag'] = 'Frequent customer'

In [19]:
ords_prods_merged['frequency_flag'].value_counts(dropna = False)

Frequent customer        20304018
Regular customer          7217134
Non-frequent customer     3639669
NaN                       1273391
Name: frequency_flag, dtype: int64

# Frequent customers are the most popular customers that purchase from the store 10 days before 

In [21]:
ords_prods_merged.to_pickle(os.path.join(path, 'ords_prods_merged.pkl'))

In [22]:
#Import data 1
ords_prods_merged = pd.read_pickle(os.path.join(path, 'ords_prods_merged.pkl'))

In [23]:
ords_prods_merged.head()

order_id  product_id  add_to_cart_order  reordered  Unnamed: 0  user_id  \
0         2       33120                  1          1     3355525   202279   
1        26       33120                  5          0     2549350   153404   
2       120       33120                 13          0      395303    23750   
3       327       33120                  5          1      976751    58707   
4       390       33120                 28          1     2764474   166654   

  eval_set  order_number  order_dow  order_hour_of_day  ...  \
0    prior             3          5                  9  ...   
1    prior             2          0                 16  ...   
2    prior            11          6                  8  ...   
3    prior            21          6                  9  ...   
4    prior            48          0                 12  ...   

     price_range_loc     busiest_day    busiest_days  busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy         Average orders   
1  Mid-range product     Busiest day    Busiest days         Average orders   
2  Mid-range product  Regularly busy  Regularly busy         Average orders   
3  Mid-range product  Regularly busy  Regularly busy         Average orders   
4  Mid-range product     Busiest day    Busiest days            Most orders   

   max_order      loyalty_flag avg_price spender_flag median_order_duration  \
0          8      New customer  8.618889  Low spender                  24.0   
1         32  Regular customer  8.072613  Low spender                   8.0   
2         13  Regular customer  8.598592  Low spender                  10.0   
3         30  Regular customer  6.549333  Low spender                  10.0   
4         62    Loyal customer  7.302487  Low spender                   5.0   

          frequency_flag  
0  Non-frequent customer  
1      Frequent customer  
2                    NaN  
3                    NaN  
4      Frequent customer  

[5 rows x 27 columns]